### PR TITLE
Heterogeneous Graph Tutorial

### PR DESCRIPTION
**Pull request by rusty1s**
Fixes https://github.com/vrtex-team/pytorch_geometric/issues/23, https://github.com/vrtex-team/pytorch_geometric/issues/67, https://github.com/vrtex-team/pytorch_geometric/issues/79

### CODE REVIEW

**Review by mrjel — APPROVED** [review]

File: docs/source/notes/heterogeneous.rst [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+============================
+
+* What is a heterogeneous graph?
+* What are the use-cases of learning on heterogeneous graphs?
+* Figure: For example, from OGB-LSC mag240m graph
+
+Creating Heterogeneous Graphs
+-----------------------------
+
+* :class:`torch_geometric.data.HeteroData`
```
Comment (yaoyaowd) [inline docs/source/notes/heterogeneous.rst:11]: The visualization for this is a little strange.

File: docs/source/notes/heterogeneous.rst [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+   data = T.ToUndirected()(data)
+   data = T.AddSelfLoops()(data)
+   data = T.NormalizeFeatures()(data)
+
+Creating Heterogeneous Models
+-----------------------------
+
+* Lazy initialization
+* :meth:`torch_geometric.nn.to_hetero`
```
Comment (yaoyaowd) [inline docs/source/notes/heterogeneous.rst:123]: ditto.

**Review by yaoyaowd — APPROVED** [review]

File: docs/source/notes/heterogeneous.rst (hunk 1 above)
Comment (rusty1s) [reply docs/source/notes/heterogeneous.rst:11]: Not sure I understand. Can you elaborate?

Comment (yaoyaowd) [reply docs/source/notes/heterogeneous.rst:11]: The UI looks like this:
Screen Shot 2021-08-08 at 10 14 46 PM
While I was expecting something like:
Screen Shot 2021-08-08 at 10 15 32 PM

But I'm ok with both. Not super familiar with rst file.

Comment (rusty1s) [reply docs/source/notes/heterogeneous.rst:11]: This will get properly when building the docs via:
```
cd docs
make clean && make html
open build/html/index.html
```
Screen Shot 2021-08-09 at 10 49 35

File: docs/source/notes/heterogeneous.rst [hunk 3]
```diff
@@ -0,0 +1,345 @@
+Heterogeneous Graph Learning
+============================
+
+A large set of real-world datasets are stored as heterogeneous graphs, motivating the introduction of specialized functionality for them in Pytorch Geometric.
+For example, the information from relational databases can be interpreted as such, as they store information about different types of entities and their (different) relations.
```
Comment (rusty1s) [inline docs/source/notes/heterogeneous.rst:5]: We should give a different example to not spoil our idea. Maybe just say that real-world graphs are typically heterogeneous involving a lot of different node and edge types, such as in graphs used for recommendation (Amazon product graphs, ...).

File: docs/source/notes/heterogeneous.rst [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+  :width: 500px
+
+The given heterogeneous graph has 244,160,499 nodes, split between the three node types **Author**, **Paper** and **Institution**.
+It further has 1,728,364,232 edges, which also are of one of three types:
+
+* **writes**: between Author and Paper
+* **affiliated with**: between Author and Institution
+* **cites**: between Paper and Paper
+
+Given this homogenous graph, we aim to infer the primary subject areas of the paper, which is one node feature of the Paper type nodes.
```
Comment (rusty1s) [inline docs/source/notes/heterogeneous.rst:28]: heterogeneous

File: docs/source/notes/heterogeneous.rst [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+
+   data = HeteroData()
+
+   data['paper'].x = ...  # [num_papers, num_features_paper]
+   data['author'].x = ...  # [num_authors, num_features_author]
+   data['institution'].x = ...  # [num_institutions, num_features_institution]
+
+   data['paper', 'cites', 'paper'].edge_index = ...  # [2, num_edges_cites]
+   data['author', 'writes', 'paper'].edge_index = ...  # [2, num_edges_writes]
+   data['author', 'affiliated_with', 'institution'].edge_index = ...  # [2, num_edges_affiliated]
```
Comment (rusty1s) [inline docs/source/notes/heterogeneous.rst:46]: Add `edge_attr` to make the example align with the text.

File: docs/source/notes/heterogeneous.rst [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+
+Most transformations for preprocessing regular graphs work as well on the heterogeneous graph data object:
+
+.. code-block:: python
+
+   import torch_geometric.transforms as T
+
+   data = T.ToUndirected()(data)
+   data = T.AddSelfLoops()(data)
+   data = T.NormalizeFeatures()(data)
```
Comment (rusty1s) [inline docs/source/notes/heterogeneous.rst:131]: Give more details on how these transforms work, in particular `T.ToUndirected`.

File: docs/source/notes/heterogeneous.rst [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+1) Automatically convert a homogenous MP-GNN to a heterogeneous MP-GNN model using :meth:`torch_geometric.nn.to_hetero`,
+2) define inidividual functions for different types using PyGs wrapper for heterogeneous convolution :class:`torch_geometric.nn.conv.HeteroConv`, or 
+3) deploy existing heterogeneous MP-GNN operators.
+
+In the following, each option is introduced in detail.
+
+Automatically converting MP-GNN models
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Pytorch Geometric allows to automatically convert any PyG MP-GNN model to a model for heterogeneous input graphs, using the built in function :meth:`torch_geometric.nn.to_hetero`.
```
Comment (rusty1s) [inline docs/source/notes/heterogeneous.rst:150]: Add the `to_hetero.svg` figure here.

File: docs/source/notes/heterogeneous.rst [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+
+.. code-block:: python
+
+   model = to_hetero(model, data.metadata(), aggr='sum')
+   
+Since some modules use lazy initialization, the model needs to be initialized by calling it once.
```
Comment (rusty1s) [inline docs/source/notes/heterogeneous.rst:156]: This needs to be explained in more detail.

File: docs/source/notes/heterogeneous.rst [hunk 9]
```diff
@@ -0,0 +1,345 @@
...
+
+   edge_type_list = [('paper', 'cites', 'paper'), ('author', 'writes', 'paper'), ('author', 'affiliated_with', 'institution')]
+   class HeteroGNN(torch.nn.Module):
+      def __init__(self, metadata, hidden_channels, out_channels, num_layers):
+         super().__init__()
+
+         self.convs = torch.nn.ModuleList()
+         for _ in range(num_layers):
+               conv = HeteroConv({
+                  edge_type: SAGEConv(-1, hidden_channels)
```
Comment (rusty1s) [inline docs/source/notes/heterogeneous.rst:205]: `HeteroConv` has the advantage of allowing different GNN ops for different edge types (we might want to show-case that here).

File: docs/source/notes/heterogeneous.rst [hunk 10]
```diff
@@ -0,0 +1,345 @@
...
+         out = model(data.x_dict, data.edge_index_dict)
+
+Afterwards, the created model can be trained using heterogeneous graph data.
+
+.. code-block:: python
+
+   def train():
+      model.train()
+      optimizer.zero_grad()
+      out = model(data.x_dict, data.edge_index_dict)
```
Comment (rusty1s) [inline docs/source/notes/heterogeneous.rst:170]: Some notes on `x_dict` and `edge_index_dict`.

File: docs/source/notes/heterogeneous.rst [hunk 11]
```diff
@@ -0,0 +1,345 @@
...
+               for edge_type in edge_type_list
+            },
+            # Use a batch size of 128 for sampling training nodes of type paper
+            batch_size=128,
+            input_nodes: ('paper', data['paper'].train_mask),
+        )
+        sampled_hetero_data = next(iter(loader))
+
+In contrast to sampling on homogenous graphs, we simply provide the heterogeneous data object and a dictionary of lists instead of a single list, containing the number of neighbors to sample in each iteration.
+Using the input_nodes parameter, we specify the type and indices of nodes, from which we want to sample.
```
Comment (rusty1s) [inline docs/source/notes/heterogeneous.rst:341]: :obj:`input_nodes`

**Review by rusty1s — COMMENTED** [review]
Building the documentation fails due to:
```
Warning, treated as error:
/home/runner/work/pytorch_geometric/pytorch_geometric/docs/source/notes/heterogeneous.rst:183:unknown document: ../../../examples/hetero/hetero_conv_dblp.py
```

File: docs/source/notes/heterogeneous.rst (hunk 8 above)
Comment (mrjel) [reply docs/source/notes/heterogeneous.rst:156]: Are you referring to the lazy initialization or the to_hetero()?

Comment (rusty1s) [reply docs/source/notes/heterogeneous.rst:156]: Lazy initialization and why it's necessary (i.e. features of nodes across different types can have different feature dimensionalities, and so lazy initialization is an elegant way to avoid keeping track of this).

File: docs/source/notes/heterogeneous.rst (hunk 10 above)
Comment (mrjel) [reply docs/source/notes/heterogeneous.rst:178]: Added that part to the introduction of the data object in the beginning.

File: docs/source/notes/heterogeneous.rst [hunk 12]
```diff
@@ -0,0 +1,375 @@
...
+Example
+----------------------------
+As a guiding example, we take a look at the heterogenous `OGB-LSC mag240m graph <https://ogb.stanford.edu/kddcup2021/mag240m/>`_ from the `OGB datasets <https://ogb.stanford.edu/>`_.
+
+.. image:: ../_figures/hg_example.svg
+  :align: center
+  :width: 500px
+
+The given heterogeneous graph has 244,160,499 nodes, split between the three node types **Author**, **Paper** and **Institution**.
+It further has 1,728,364,232 edges, which also are of one of three types:
```
Comment (JiaxuanYou) [inline docs/source/notes/heterogeneous.rst:22]: Are we using the OGB 240M dataset in the example later?
HeteroData(
      paper={ x=[736389, 128] },
      author={ x=[1134649, 64] },
      institution={ x=[8740, 256] },
      ('paper', 'cites', 'paper')={ edge_index=[2, 5416271] },
      ('author', 'writes', 'paper')={ edge_index=[2, 7145660] },
      ('author', 'affiliated_with', 'institution')={ edge_index=[2, 1043998] }
   )
which seems to be a smaller dataset. I assume it's a subset of OGB?

**Review by JiaxuanYou — APPROVED** [review]
Great job! Just had one comment on the dataset selection

File: docs/source/notes/heterogeneous.rst [hunk 13]
```diff
@@ -0,0 +1,375 @@
...
+        from torch_geometric.datasets import OGB_MAG
+        from torch_geometric.loader import NeighborLoader
+        hetero_data = OGB_MAG(path)[0]
+        loader = NeighborLoader(
+            hetero_data,
+            # Sample 30 neighbors for each node and iteration in each edge type
+            # for 2 iterations
+            num_neighbors={
+            edge_type: [30] * 2
+               for edge_type in edge_type_list
```
Comment (rusty1s) [inline docs/source/notes/heterogeneous.rst:362]:
```suggestion
               for edge_type in hetero_data.edge_types
```

File: docs/source/notes/heterogeneous.rst (hunk 12 above)
Comment (mrjel) [reply docs/source/notes/heterogeneous.rst]: Thanks for pointing that out.
I changed everything to consistently use the OGB-MAG dataset now. It is smaller and probably better suited for the example.
